File: sidecar/src/diagnostics/InteractionLog.ts
import fs from "node:fs"
import os from "node:os"
import path from "node:path"

const MAX_BYTES = 8 * 1024 * 1024
const MAX_LINE_CHARS = 96 * 1024
const MAX_STRING_CHARS = 4096
const MAX_ARRAY_ITEMS = 50
const MAX_OBJECT_KEYS = 80
const MAX_DEPTH = 8
const VERBOSE_INTERACTION_LOG = process.env.VSCLINE_VERBOSE_INTERACTION_LOG === "1"
const ENABLE_INTERACTION_LOG = VERBOSE_INTERACTION_LOG || process.env.VSCLINE_ENABLE_INTERACTION_LOG === "1"

const SENSITIVE_KEYS = [
	"apikey",
	"api_key",
	"authorization",
	"password",
	"token",
	"secret",
	"cookie",
]

export function logInteraction(direction: string, event: string, payload?: unknown) {
	try {
		if (!ENABLE_INTERACTION_LOG && !isImportantDiagnosticEvent(event)) {
			return
		}
		if (shouldSkipDefaultLog(direction, event, payload)) {
			return
		}
		const entry = {
			at: new Date().toISOString(),
			source: "sidecar",
			direction,
			event,
			payload: sanitize(compactPayload(direction, event, payload), 0),
		}
		let line = JSON.stringify(entry)
		if (line.length > MAX_LINE_CHARS) {
			line = `${line.slice(0, MAX_LINE_CHARS)}...[truncated]`
		}
		const filePath = getLogPath()
		fs.mkdirSync(path.dirname(filePath), { recursive: true })
		rotateIfNeeded(filePath)
		fs.appendFileSync(filePath, `${line}\n`, "utf8")
	} catch {
		// Diagnostics must never interfere with the extension.
	}
}

function isImportantDiagnosticEvent(event: string) {
	const normalized = event.toLowerCase()
	return normalized.includes("failed") || normalized.includes("error") || normalized.includes("slow")
}

function shouldSkipDefaultLog(direction: string, event: string, payload: unknown) {
	if (VERBOSE_INTERACTION_LOG) {
		return false
	}

	if (event === "jsonrpc.line" || event === "jsonrpc.response.result" || event === "partialMessage" || event === "taskActivity") {
		return true
	}

	if (event === "webview.postMessage" || event === "webview.message.batchItem") {
		return true
	}

	if (event.endsWith(".result") && isPostedWebviewResult(payload)) {
		return true
	}

	if (event === "sdk.event" && isHighFrequencySdkEvent(payload)) {
		return true
	}

	return false
}

function compactPayload(direction: string, event: string, payload: unknown) {
	if (event === "webview.message" && typeof payload === "string") {
		return summarizeWebviewMessage(payload)
	}

	if (event === "state.broadcast") {
		const record = asRecord(payload)
		const messages = Array.isArray(record.messages) ? record.messages : []
		return {
			count: record.count,
			messages: messages.slice(0, 4),
			truncatedMessages: Math.max(0, messages.length - 4),
		}
	}

	if (direction === "host->sidecar" && event === "webview.message.result") {
		const record = asRecord(payload)
		const webviewMessages = Array.isArray(record.webviewMessages) ? record.webviewMessages : []
		return {
			handled: record.handled,
			webviewMessageCount: webviewMessages.length,
		}
	}

	return payload
}

function summarizeWebviewMessage(rawJson: string) {
	const parsed = tryParseJson(rawJson)
	const record = asRecord(parsed)
	const request = asRecord(record.grpc_request)
	const cancel = asRecord(record.grpc_request_cancel)
	return {
		type: getString(record, "type"),
		service: getString(request, "service"),
		method: getString(request, "method"),
		requestId: getString(request, "request_id") || getString(request, "requestId") || getString(cancel, "request_id"),
		isStreaming: request.is_streaming === true || request.isStreaming === true,
		rawLength: rawJson.length,
	}
}

function isPostedWebviewResult(payload: unknown) {
	const result = asRecord(asRecord(payload).result)
	return result.posted === true && Object.keys(result).length <= 1
}

function isHighFrequencySdkEvent(payload: unknown) {
	const record = asRecord(payload)
	if (getString(record, "type") !== "agent_event") {
		return false
	}

	const event = asRecord(record.event)
	const type = getString(event, "type")
	const contentType = getString(event, "contentType")
	if ((type === "content_update" || type === "content_delta") && (contentType === "text" || contentType === "reasoning")) {
		return true
	}
	return type === "content_start" ||
		type === "content_end" ||
		type === "iteration_start" ||
		type === "iteration_end" ||
		type === "usage"
}

function getLogPath() {
	const localAppData = process.env.LOCALAPPDATA || path.join(os.homedir(), "AppData", "Local")
	return path.join(localAppData, "VsClineAgent", "logs", `interaction-${dateStamp()}.jsonl`)
}

function dateStamp() {
	const now = new Date()
	const year = now.getFullYear()
	const month = String(now.getMonth() + 1).padStart(2, "0")
	const day = String(now.getDate()).padStart(2, "0")
	return `${year}${month}${day}`
}

function rotateIfNeeded(filePath: string) {
	if (!fs.existsSync(filePath)) {
		return
	}
	const stat = fs.statSync(filePath)
	if (stat.size < MAX_BYTES) {
		return
	}
	const archive = `${filePath}.1`
	if (fs.existsSync(archive)) {
		fs.rmSync(archive, { force: true })
	}
	fs.renameSync(filePath, archive)
}

function sanitize(value: unknown, depth: number): unknown {
	if (value === null || value === undefined) {
		return value
	}
	if (typeof value === "string") {
		const parsed = tryParseJson(value)
		return parsed === undefined ? truncateDiagnosticString(redactSecretLikeString(value)) : sanitize(parsed, depth + 1)
	}
	if (typeof value !== "object") {
		return value
	}
	if (depth >= MAX_DEPTH) {
		return "[max-depth]"
	}
	if (Array.isArray(value)) {
		const items = value.slice(0, MAX_ARRAY_ITEMS).map((item) => sanitize(item, depth + 1))
		return value.length > MAX_ARRAY_ITEMS ? [...items, `[truncated ${value.length - MAX_ARRAY_ITEMS} items]`] : items
	}

	const result: Record<string, unknown> = {}
	const entries = Object.entries(value as Record<string, unknown>)
	for (const [key, nested] of entries.slice(0, MAX_OBJECT_KEYS)) {
		result[key] = isSensitiveKey(key) ? redactValue(nested) : sanitize(nested, depth + 1)
	}
	if (entries.length > MAX_OBJECT_KEYS) {
		result.__truncatedKeys = entries.length - MAX_OBJECT_KEYS
	}
	return result
}

function asRecord(value: unknown): Record<string, unknown> {
	return value !== null && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {}
}

function getString(record: Record<string, unknown>, key: string) {
	const value = record[key]
	return typeof value === "string" ? value : ""
}

function tryParseJson(value: string) {
	try {
		return JSON.parse(value) as unknown
	} catch {
		return undefined
	}
}

function isSensitiveKey(key: string) {
	const normalized = key.replace(/[-_\s]/g, "").toLowerCase()
	return SENSITIVE_KEYS.some((sensitive) => normalized.includes(sensitive.replace(/[-_\s]/g, ""))) || normalized === "key"
}

function redactValue(value: unknown) {
	const text = typeof value === "string" ? value : ""
	if (text.length <= 8) {
		return "[redacted]"
	}
	return `${text.slice(0, 4)}...${text.slice(-4)}`
}

function redactSecretLikeString(value: string) {
	return value.replace(
		/\b(sk-[A-Za-z0-9_-]{12,}|sk-proj-[A-Za-z0-9_-]{12,}|github_pat_[A-Za-z0-9_]{12,}|nvapi-[A-Za-z0-9_-]{12,})\b/g,
		(match) => `${match.slice(0, 7)}...${match.slice(-4)}`,
	)
}

function truncateDiagnosticString(value: string) {
	if (value.length <= MAX_STRING_CHARS) {
		return value
	}
	return `${value.slice(0, MAX_STRING_CHARS)}...[truncated ${value.length - MAX_STRING_CHARS} chars]`
}

File: src/extension/Host/InteractionLog.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace VsClineAgent.Host
{
    internal static class InteractionLog
    {
        private const long MaxBytes = 8L * 1024L * 1024L;
        private const int MaxLineChars = 96 * 1024;
        private const long MaxTotalBytes = 150L * 1024L * 1024L;
        private static readonly object Gate = new object();
        private static string _lastCleanupDate = "";

        public static void Write(string direction, string eventName, object? payload)
        {
            try
            {
                if (ShouldSkipDefaultLog(eventName, payload))
                    return;

                var compactPayload = CompactPayload(eventName, payload);
                var entry = new JObject
                {
                    ["at"] = DateTimeOffset.Now.ToString("O"),
                    ["source"] = "vsix-host",
                    ["direction"] = direction,
                    ["event"] = eventName,
                    ["payload"] = InteractionLogSanitizer.Sanitize(compactPayload)
                };
                var correlationId = InteractionCorrelation.FromPayload(compactPayload);
                if (!string.IsNullOrWhiteSpace(correlationId))
                    entry["correlationId"] = correlationId;

                var line = entry.ToString(Formatting.None);
                if (line.Length > MaxLineChars)
                    line = line.Substring(0, MaxLineChars) + "...[truncated]";

                lock (Gate)
                {
                    var path = GetLogPath();
                    RotateIfNeeded(path);
                    File.AppendAllText(path, line + Environment.NewLine, Encoding.UTF8);
                }
            }
            catch
            {
            }
        }

        private static bool ShouldSkipDefaultLog(string eventName, object? payload)
        {
            if (string.Equals(Environment.GetEnvironmentVariable("VSCLINE_VERBOSE_INTERACTION_LOG"), "1", StringComparison.Ordinal))
                return false;

            if (!string.Equals(Environment.GetEnvironmentVariable("VSCLINE_ENABLE_INTERACTION_LOG"), "1", StringComparison.Ordinal)
                && eventName.IndexOf("failed", StringComparison.OrdinalIgnoreCase) < 0
                && eventName.IndexOf("error", StringComparison.OrdinalIgnoreCase) < 0
                && eventName.IndexOf("slow", StringComparison.OrdinalIgnoreCase) < 0)
                return true;

            if (eventName == "webview.postMessage" || eventName == "webview.message.batchItem")
                return true;

            if (eventName == "webview.message.result")
            {
                var token = payload as JToken;
                var webviewMessages = token?["webviewMessages"] as JArray;
                return webviewMessages != null && webviewMessages.Count > 0;
            }

            return false;
        }

        private static object? CompactPayload(string eventName, object? payload)
        {
            if (eventName == "webview.message" && payload is string rawJson)
                return SummarizeWebviewMessage(rawJson);

            return payload;
        }

        private static JObject SummarizeWebviewMessage(string rawJson)
        {
            var parsed = TryParseJson(rawJson) as JObject;
            var request = parsed?["grpc_request"] as JObject;
            var cancel = parsed?["grpc_request_cancel"] as JObject;
            return new JObject
            {
                ["type"] = parsed?["type"]?.ToString(),
                ["service"] = request?["service"]?.ToString(),
                ["method"] = request?["method"]?.ToString(),
                ["requestId"] = request?["request_id"]?.ToString() ?? request?["requestId"]?.ToString() ?? cancel?["request_id"]?.ToString(),
                ["isStreaming"] = request?["is_streaming"]?.Value<bool?>() ?? request?["isStreaming"]?.Value<bool?>() ?? false,
                ["rawLength"] = rawJson.Length
            };
        }

        private static JToken? TryParseJson(string text)
        {
            try
            {
                return JToken.Parse(text);
            }
            catch
            {
                return null;
            }
        }

        private static string GetLogPath()
        {
            var directory = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "VsClineAgent",
                "logs");
            Directory.CreateDirectory(directory);
            CleanupLogsOncePerDay(directory);
            return Path.Combine(directory, "interaction-" + DateTime.Now.ToString("yyyyMMdd") + ".jsonl");
        }

        private static void CleanupLogsOncePerDay(string directory)
        {
            var date = DateTime.Now.ToString("yyyyMMdd");
            if (string.Equals(_lastCleanupDate, date, StringComparison.Ordinal))
                return;
            _lastCleanupDate = date;
            var retentionDays = string.Equals(Environment.GetEnvironmentVariable("VSCLINE_VERBOSE_INTERACTION_LOG"), "1", StringComparison.Ordinal) ? 7 : 14;
            var currentPath = Path.GetFullPath(Path.Combine(directory, "interaction-" + date + ".jsonl"));
            try
            {
                var cutoff = DateTime.UtcNow.AddDays(-retentionDays);
                var files = new DirectoryInfo(directory).EnumerateFiles("interaction-*.jsonl*")
                    .OrderBy(file => file.LastWriteTimeUtc).ToList();
                foreach (var file in files.Where(file => !string.Equals(file.FullName, currentPath, StringComparison.OrdinalIgnoreCase) && file.LastWriteTimeUtc < cutoff).ToList())
                    TryDelete(file);
                files = files.Where(file => file.Exists).OrderBy(file => file.LastWriteTimeUtc).ToList();
                var total = files.Sum(file => file.Length);
                foreach (var file in files)
                {
                    if (total <= MaxTotalBytes)
                        break;
                    if (string.Equals(file.FullName, currentPath, StringComparison.OrdinalIgnoreCase))
                        continue;
                    var length = file.Length;
                    if (TryDelete(file))
                        total -= length;
                }
            }
            catch
            {
            }
        }

        private static bool TryDelete(FileInfo file)
        {
            try { file.Delete(); return true; }
            catch { return false; }
        }

        private static void RotateIfNeeded(string path)
        {
            if (!File.Exists(path))
                return;

            var info = new FileInfo(path);
            if (info.Length < MaxBytes)
                return;

            var archive = path + ".1";
            if (File.Exists(archive))
                File.Delete(archive);
            File.Move(path, archive);
        }
    }
}
